webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala

```diff
@@ -90,4 +96,95 @@ class ReadOnlySoftwareDAOImpl(
       softwareByNode.mapValues { ids => software.filter(s => ids.contains(s.id)) }
     }
   }
+
+  def getAllSoftwareIds() : Box[Set[SoftwareUuid]] = {
+    for {
+      con     <- ldap
+      softwareEntry = con.searchOne(inventoryDitService.getSoftwareBaseDN, IS(OC_SOFTWARE), A_SOFTWARE_UUID)
+      ids     <- sequence(softwareEntry) { entry =>
+        entry(A_SOFTWARE_UUID) match {
+          case Some(value) => Full(value)
+          case _ => Failure(s"Missing attribute ${A_SOFTWARE_UUID} for entry ${entry.dn} ${entry.toString()}")
+        }
+      }
+      softIds       =  ids.map(id => SoftwareUuid(id)).toSet
+    } yield {
+      softIds
+    }
+  }
+
+  def getSoftwaresForAllNodes() : Box[Set[SoftwareUuid]] = {
+    // fetch all softwares, for all nodes, in all 3 dits
+    val acceptedDit = inventoryDitService.getDit(AcceptedInventory)
+
+    var mutSetSoftwares: Box[scala.collection.mutable.Set[SoftwareUuid]] = Full(scala.collection.mutable.Set[SoftwareUuid]())
+
+    val t1 = System.currentTimeMillis
+    for {
+      con           <- ldap
+      // fetch all nodes
+      nodes         = con.searchSub(acceptedDit.NODES.dn.getParent, IS(OC_NODE), A_NODE_UUID)
+
+      batchedNodes = nodes.grouped(50)
+
+      _ = batchedNodes.foreach { nodeEntries: Seq[LDAPEntry] =>
+                             val nodeIds = nodeEntries.flatMap(_(A_NODE_UUID)).map(NodeId(_))
+
+                             val orFilter: Filter = BuildFilter.OR(nodeIds.map(x => EQ(A_NODE_UUID, x.value)): _*)
+                             val softwareEntry: Seq[LDAPEntry] =  con.searchSub(acceptedDit.NODES.dn.getParent, orFilter, A_SOFTWARE_DN)
+                             val ids: Seq[String] = softwareEntry.flatMap(entry => entry.valuesFor(A_SOFTWARE_DN).toSet )
+                             val results = sequence(ids) { id => acceptedDit.SOFTWARE.SOFT.idFromDN(new DN(id)) }
+
+                             results match {
+                               case Full(softIds) =>
+                                 mutSetSoftwares = mutSetSoftwares.map(t => t ++ softIds)
+                                 Full(Unit)
+                               case Failure(msg, exception, chain) => mutSetSoftwares = Failure(msg, exception, chain) // otherwise the time is wrong
+                             }
+                           }
```

Comment:
this ain't pretty, yet its more efficient
[2019-05-23 13:00:06] DEBUG com.normation.inventory.ldap.core.SoftwareServiceImpl - All softwares id in nodes fetched: 37953 softwares id in 10029ms, compared to what is in the TODO below
[2019-05-23 12:55:17] DEBUG com.normation.inventory.ldap.core.SoftwareServiceImpl - All softwares id in nodes fetched: 37953 softwares id in 15197ms

Reply:
this code is surely non idiomatic, and i fail to improve it